### PR TITLE
Support Fargate services

### LINKED ISSUE #5: Container instance cannot be empty
Thanks for writing this.

Getting the following error `Container instance cannot be empty`. Also, once this is working, how are we supposed to get the task definition arn as it changes every deploy? Are we supposed to manually call out to the AWS API to get the task definition arn? Seems like providing the cluster and service name should be enough.

```
npx ecs-deployment-monitor --cluster arn:aws:ecs:us-west-2:XXXXXXXXXXXX:cluster/prod --service-name api --task-definition arn:aws:ecs:us-west-2:XXXXXXXXXXXX:task-definition/api:33

npx: installed 83 in 3.419s
/Users/justin/.npm/_npx/13914/lib/node_modules/ecs-deployment-monitor/node_modules/aws-sdk/lib/request.js:31
            throw err;
            ^

InvalidParameterException: Container instance cannot be empty.
    at Request.extractError (/Users/justin/.npm/_npx/13914/lib/node_modules/ecs-deployment-monitor/node_modules/aws-sdk/lib/protocol/json.js:51:27)
    at Request.callListeners (/Users/justin/.npm/_npx/13914/lib/node_modules/ecs-deployment-monitor/node_modules/aws-sdk/lib/sequential_executor.js:106:20)
    at Request.emit (/Users/justin/.npm/_npx/13914/lib/node_modules/ecs-deployment-monitor/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
    at Request.emit (/Users/justin/.npm/_npx/13914/lib/node_modules/ecs-deployment-monitor/node_modules/aws-sdk/lib/request.js:688:14)
    at Request.transition (/Users/justin/.npm/_npx/13914/lib/node_modules/ecs-deployment-monitor/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/Users/justin/.npm/_npx/13914/lib/node_modules/ecs-deployment-monitor/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /Users/justin/.npm/_npx/13914/lib/node_modules/ecs-deployment-monitor/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/Users/justin/.npm/_npx/13914/lib/node_modules/ecs-deployment-monitor/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/Users/justin/.npm/_npx/13914/lib/node_modules/ecs-deployment-monitor/node_modules/aws-sdk/lib/request.js:690:12)
    at Request.callListeners (/Users/justin/.npm/_npx/13914/lib/node_modules/ecs-deployment-monitor/node_modules/aws-sdk/lib/sequential_executor.js:116:18) {
  message: 'Container instance cannot be empty.',
  code: 'InvalidParameterException',
  time: 2020-07-15T20:12:44.910Z,
  requestId: '189b4fdc-8d9f-4b37-8c8b-29762ab57672',
  statusCode: 400,
  retryable: false,
  retryDelay: 72.30983724608417
}
```

### PR DESCRIPTION
**Pull request by ohookins**
Fargate support was missing - since there are no container instances backing Fargate tasks, any attempt to check the health of the container instance in the target group would fail. Instead, Fargate tasks have their own awsvpc networking and are attached to the target group by IP directly. Thus, checking the health of the target group entries is sufficient for understanding the health of the tasks themselves.

Have now added tests as well, although they are so heavily mocked and supported by fixtures that they are a bit unsatisfying. Any feedback welcome.

Closes #5

### CODE REVIEW

**Review by Liath — APPROVED** [review]
LGTM!

File: lib/service.js [hunk 1]
```diff
@@ -19,6 +19,7 @@ class Service extends EventEmitter {
     this.tasks = [];
     this.initiated = false;
     this.stop = false;
+    this.launchType = '';
```
Comment (zarakay) [inline lib/service.js:22]: default this to `EC2`?

File: test/service.js [hunk 2]
```diff
@@ -188,7 +196,7 @@ describe('Service', function() {
       containerInstanceStub.restore();
 
       AWS.mock('ECS', 'listContainerInstances', function (params, cb) {
-        expect(params.cluster).to.equal('cluster-yo');
+        expect(params.cluster).to.equal('cluster-yo6');
```
Comment (zarakay) [inline test/service.js:199]: Can we remove the `yo` bit?

**Review by zarakay — CHANGES_REQUESTED** [review]
minor changes

**Comment by ohookins** [comment]
OK, this definitely needs some more work. Right now we are just checking that the number of healthy targets listed against the target group matches the number of tasks we have in the deployment, but of course that ignores the previous deployment - so the monitor will return early. We need to first resolve the bind IPs/ports of the Fargate tasks from their container descriptions and only check those in the target group - so I guess this is very similar to how the EC2 version works.

Few more functions required and perhaps better tests but I think I understand it a lot better now.

File: lib/service.js [hunk 3]
```diff
@@ -307,28 +377,33 @@ class Service extends EventEmitter {
...
     if (!task) {
       this.emit('warning', `Task ${taskArn} does not exist. Or has not been cached against the service`);
       return false;
     }
 
-    var containerInstance = this.getContainerInstance(task.containerInstanceArn);
+    // Most of this logic is identical for Fargate- and EC2-backed containers,
+    // except for the inner logic around finding the targets.
+    const isHealthy = this.raw.loadBalancers.every((lb) => {
+      const container = task.containers.find((container) => container.name === lb.containerName);
```
Comment (zarakay) [inline lib/service.js:389]: Do we need to check the value of `container`?

File: lib/service.js [hunk 4]
```diff
@@ -213,6 +218,52 @@ class Service extends EventEmitter {
...
+   * @param {object} container The specific container we are looking for targets associated with.
+   * @return {array} A list of targets
+   */
+  _getTargetsForEC2Task(task, container) {
+    const containerInstance = this.getContainerInstance(task.containerInstanceArn);
+
+    return container.networkBindings.map((binding) => {
+      return this.getTarget(containerInstance.ec2InstanceId, binding.hostPort);
+    }).filter((target) => !!target);
+  }
```
Comment (zarakay) [inline lib/service.js:236]: Random Aside: If a task is running in awsvpc mode on EC2, I assume this still works?

File: lib/service.js [hunk 5]
```diff
@@ -213,6 +218,52 @@ class Service extends EventEmitter {
...
+
+  /**
+   * Get Targets For Fargate Task
+   *
+   * Finds the load balancer targets for the Farget task based on elastic network
```
Comment (zarakay) [inline lib/service.js:241]:
```suggestion
   * Finds the load balancer targets for the Fargate task based on elastic network
```

File: lib/service.js (hunk 4 above)
Comment (ohookins) [reply lib/service.js:236]: I have no idea. I'd prefer to make that change (if we deem it necessary) later on as a separate PR.

File: lib/service.js (hunk 3 above)
Comment (ohookins) [reply lib/service.js:389]: I guess the failure condition here would be that we have a task but has no containers defined. Is that even legal?
Previously all of the EC2-based logic had an implicit assumption that the container would be found so it was never checked. I can add an explicit failure condition here though.

**Review by zarakay — APPROVED** [review]

**Pull request merged**